File: bayesmendel-comparator/r/tests/testthat/test-risk_api_client.R
PEDIGREE_ID <- "11111111-1111-1111-1111-111111111111"
COUNSELEE_ID <- "22222222-2222-2222-2222-222222222222"

recording_gateway <- function(status, json_body) {
  last <- new.env(parent = emptyenv())
  last$url <- NULL
  last$headers <- NULL
  last$body_json <- NULL

  gateway <- function(request) {
    last$url <- request$url
    last$headers <- request$headers
    last$body_json <- request$body_json
    list(status = status, json_body = json_body)
  }
  list(gateway = gateway, last = last)
}

test_that("posts the requested model to risk/calculate", {
  for (model in c("BRCAPRO", "MMRpro", "PancPRO")) {
    recorder <- recording_gateway(
      status = 200L,
      json_body = '{"carrier_probabilities": {}, "future_risks": []}'
    )

    calculate_risk(
      recorder$gateway,
      base_url = "https://evagene.example",
      api_key = "evg_test",
      pedigree_id = PEDIGREE_ID,
      model = model
    )

    expect_equal(
      recorder$last$url,
      sprintf("https://evagene.example/api/pedigrees/%s/risk/calculate", PEDIGREE_ID)
    )
    expect_equal(unname(recorder$last$headers[["X-API-Key"]]), "evg_test")

    body <- jsonlite::fromJSON(recorder$last$body_json, simplifyVector = FALSE)
    expect_equal(body$model, model)
    expect_null(body$counselee_id)
  }
})

test_that("includes counselee_id when provided", {
  recorder <- recording_gateway(status = 200L, json_body = '{}')

  calculate_risk(
    recorder$gateway,
    base_url = "https://evagene.example",
    api_key = "evg_test",
    pedigree_id = PEDIGREE_ID,
    model = "BRCAPRO",
    counselee_id = COUNSELEE_ID
  )

  body <- jsonlite::fromJSON(recorder$last$body_json, simplifyVector = FALSE)
  expect_equal(body$counselee_id, COUNSELEE_ID)
})

test_that("raises api_error on non-2xx status", {
  recorder <- recording_gateway(status = 500L, json_body = '{}')

  expect_error(
    calculate_risk(
      recorder$gateway, "https://evagene.example", "evg_test",
      PEDIGREE_ID, "BRCAPRO"
    ),
    class = "api_error"
  )
})

test_that("raises api_error on non-object JSON", {
  recorder <- recording_gateway(status = 200L, json_body = '["not","an","object"]')

  expect_error(
    calculate_risk(
      recorder$gateway, "https://evagene.example", "evg_test",
      PEDIGREE_ID, "BRCAPRO"
    ),
    class = "api_error"
  )
})

File: bayesmendel-comparator/python/tests/test_risk_api_client.py
from typing import Any

import pytest

from bayesmendel_comparator.risk_api_client import ApiError, RiskApiClient

_PEDIGREE_ID = "11111111-1111-1111-1111-111111111111"
_COUNSELEE_ID = "22222222-2222-2222-2222-222222222222"


class _StubResponse:
    def __init__(self, status_code: int, payload: Any) -> None:
        self.status_code = status_code
        self._payload = payload

    def json(self) -> Any:
        return self._payload


class _RecordingGateway:
    def __init__(self, response: _StubResponse) -> None:
        self._response = response
        self.last_url: str = ""
        self.last_headers: dict[str, str] = {}
        self.last_body: dict[str, Any] = {}

    def post_json(
        self,
        url: str,
        *,
        headers: dict[str, str],
        body: dict[str, Any],
    ) -> _StubResponse:
        self.last_url = url
        self.last_headers = headers
        self.last_body = body
        return self._response


def _client(gateway: _RecordingGateway) -> RiskApiClient:
    return RiskApiClient(base_url="https://evagene.example", api_key="evg_test", http=gateway)


@pytest.mark.parametrize("model", ["BRCAPRO", "MMRpro", "PancPRO"])
def test_posts_requested_model_to_risk_calculate(model: str) -> None:
    gateway = _RecordingGateway(
        _StubResponse(200, {"carrier_probabilities": {}, "future_risks": []}),
    )

    _client(gateway).calculate(_PEDIGREE_ID, model)

    assert gateway.last_url == (
        f"https://evagene.example/api/pedigrees/{_PEDIGREE_ID}/risk/calculate"
    )
    assert gateway.last_headers["X-API-Key"] == "evg_test"
    assert gateway.last_body == {"model": model}


def test_includes_counselee_when_provided() -> None:
    gateway = _RecordingGateway(_StubResponse(200, {}))

    _client(gateway).calculate(_PEDIGREE_ID, "BRCAPRO", counselee_id=_COUNSELEE_ID)

    assert gateway.last_body["counselee_id"] == _COUNSELEE_ID


def test_raises_api_error_on_non_2xx() -> None:
    gateway = _RecordingGateway(_StubResponse(500, {}))

    with pytest.raises(ApiError):
        _client(gateway).calculate(_PEDIGREE_ID, "BRCAPRO")


def test_raises_api_error_on_non_object_payload() -> None:
    gateway = _RecordingGateway(_StubResponse(200, ["not", "an", "object"]))

    with pytest.raises(ApiError):
        _client(gateway).calculate(_PEDIGREE_ID, "BRCAPRO")
